feat(blog): Add article about ESP32 Mosquitto port (#483)
espressif/developer-portal: content/blog/2025/05/esp-idf-mosquitto-port/index.md

1. Processes incoming client connections
2. Handles socket read/write events
3. Manages client keepalive tracking
4. Processes message queues and subscriptions
5. Handles disconnections and cleanup

### ESP-TLS Integration
The port includes a custom implementation (`net__esp_tls.c`) that integrates ESP-TLS for secure connections:
- Handles TLS context initialization
- Manages secure socket connections
- Processes TLS handshakes and secure data transfer
- Maintains a fixed array of TLS contexts for connected clients

### Simplified message handling
```mermaid
graph TD
    A[Client Connection Request] --> B[Socket Accept]
    B --> C{TLS Enabled?}
    C -->|Yes| D[ESP-TLS Context Creation]
    C -->|No| E[Plain TCP Connection]
    D --> F[TLS Handshake]
    F --> G[Connection Ready]
    E --> G
    G --> H[Socket Multiplexing]

    subgraph "Event Loop"
        H --> I[Poll for Events]
        I --> J[Process Socket Events]
        J --> K[Handle Client Messages]
        K --> L[Route Messages to Subscribers]
        L --> I
    end

    M[Client Disconnect] --> N[Socket Cleanup]
    N --> O[Release Resources]
```

## Getting Started

> To use the ESP32 Mosquitto Port in your projects, you'll need to initialize the broker with appropriate configuration settings and provide sufficient stack space for the running task.

Reviewer: I'm missing the step of adding the component from the component manager.

Author: Have added, thanks!

---

> title: "ESP-IDF Mosquitto Port: A Lightweight MQTT Broker for ESP32"

Author: yes, "port" is traditionally used as something like "sw-layer-adaptation", same as here.

but "Port of Mosquitto" sounds weird, maybe just `"Mosquitto: Eclipse Mosquitto&trade ported to ESP-IDF"`?

> title: "ESP-IDF Mosquitto Port: A Lightweight MQTT Broker for ESP32"

Reviewer 2: According to Proper Usage of Eclipse Foundation Trademarks:

> You may not include an Eclipse Foundation Trademark in the name of a publication, certification program, domain name or book.

I like your updated title, but we don't need to repeat Mosquitto twice.

How about "Eclipse Mosquitto ported to ESP-IDF"?

Please take my suggestions into account and come up with a final name that you like most.

> title: "ESP-IDF Mosquitto Port: A Lightweight MQTT Broker for ESP32"

Author: I actually swapped the parts again and updated the title to:
```
Lightweight MQTT Broker for ESP32: Mosquitto ported to ESP-IDF
```
this starts with the important thing first (MQTT broker), says that it's Mosquitto later, and won't use any trademarks, WYT?

date: 2025-05-28
showAuthor: false
summary: "Mosquitto -- the industry-standard MQTT broker -- has been ported to ESP-IDF. Its lightweight version retains Mosquitto's core functionality and security features to run on resource-constrained IoT devices. This MQTT broker is ideal for edge computing, testing, and standalone IoT deployments. In this article, we will do an overview and show you how to get started."
authors:

>   - david-cermak

Reviewer 2: I see that you only included your name in the article card.

Do you also want to add more information?

- You can upload a picture. If you choose not to do it, please add the Espressif logo as a placeholder in `data/authors/david-cermak.json`
  ```yaml
  "image" : "img/authors/espressif.png"
  ```
- You can add your social media accounts: GItHub, LinkedIn, etc.

tags: [  "ESP-IDF", "MQTT", "Broker", "IoT", "ESP32"]
---

## Introduction

> Traditionally, MQTT brokers live in the cloud or on powerful servers - but what if we could run them directly on IoT devices? The ESP32 Mosquitto Port makes this possible, bringing broker capabilities to edge devices and enabling autonomous IoT networks that don't depend on external services.

Author: `ESP-IDF Mosquitto port` is more "correct" (as it's only the ESP-IDF SDK that supports it for now), but `"ESP32 Mosquitto port"` sounds more natural, I'll stick with the latter.

The lightweight implementation of the popular Eclipse Mosquitto&trade; broker makes this possible. Ported to ESP-IDF, this version brings MQTT broker functionality to edge devices and reduces the need for always-on external infrastructure.

Optimized for the limited resources of the ESP32, this port maintains the core features of Eclipse Mosquitto and benefits from its feature completeness, configurability, extensibility, and portability. Running this MQTT broker directly on the device reduces latency, enhances privacy by keeping data local, and allows for fully autonomous IoT networks -- even in the absence of internet connectivity.

> ### Key Features

Reviewer 2: Throughout the article, I see that you use sentence-style capitalization and heading-style capitalization in headings alternatively. Please choose one and follow it in all article headings.

Please take into account that we encourage the use of sentence-style capitalization on the Developer Portal:

```suggestion
### Key features
```

If you would like to know why, please let me know, and I will rpovide our reasons. We plan to include them in the Contribution guide very soon.

- Supports TCP and TLS transport (based on ESP-TLS)

> - Single listener configuration

Reviewer 2: > based on ESP-TLS

This asks for a link to ESP-TLS.

> Single listener configuration

It is not mentioned anywhere else in the article.

Suggest that you either explain it here or provide more details later in the article and make the connection more explicit.

Author: added an explanation in parentheses (same as above)

> - Configurable through a simple API

Reviewer 2: > - Supports TCP and TLS transport (based on ESP-TLS)
> - Single listener configuration

These are legit technical features. Everything is fine.

> - Configurable through a simple API`

This not a feature in itself, it is more like a convenience remark.

Suggest removing this feature.

Author: removed (and rephrased a bit to still keep the three bullets)

> ## MQTT Broker on low end IoT device

Reviewer 2: > low end

It seems not very intuitive to me what _low-end_ or _high-end_ might mean when talking about IoT devices.

Am I correct, that by _low end IoT device_ you mean they have limited resources?

If yes, I suggest not mentioning it here, because it has been said in the introduction.

How about this title instead:

```suggestion
## Use cases of running an MQTT broker on IoT devices
```

> ## MQTT Broker on low end IoT device

Reviewer 2: Previously, you mostly talk about running the MQTT broker on `ESP32 devices`. Here, you shift to `IoT devices`.

Can this MQTT broker be run on Espressif SoC series other than ESP32?

Please choose your answer and keep it consistent throughout the article.

> While it is possible to run an MQTT broker on ESP32 devices, what are the use cases and considerations? Let's explore three interesting scenarios where running the broker on ESP32 is particularly beneficial:

Reviewer 2: ```suggestion
While it is possible to run an MQTT broker on ESP32 devices, what are the use cases and considerations? Let's explore three interesting scenarios where running the broker on ESP32 is particularly beneficial.
```

### Local Broker for Private IoT networks
One powerful use case is running a local Mosquitto broker on the ESP32 to serve local IoT devices within a private network segment:
- Create an independent home automation hub that works even when internet connectivity fails
- Integrate with platforms like ESP-Home, Home Assistant, or custom IoT devices
- Enhance privacy by keeping all data within the local network

### Testing on Target
Another use-case is testing MQTT applications directly on the ESP32. By running both the client and broker on the same chip, developers can:
- Develop and test MQTT-based applications without external dependencies
- Simulate complex MQTT interactions in a controlled environment
- Debug communication issues in isolation before deploying to a production environment
- Create self-contained test fixtures for IoT product validation

### Bridged Brokers over P2P Networks

> The serverless_mqtt example demonstrates how two ESP32 devices running Mosquitto can create a synchronized broker system over peer-to-peer connections:

Reviewer 2: How about providing a link to the serverless_mqtt example?

Author: link added

- Connect IoT devices across separate private networks without requiring cloud infrastructure
- Synchronize two local MQTT networks across the internet using secure peer-to-peer connections
- Create resilient IoT networks that can operate across geographical locations
- Enable MQTT communication between networks separated by NATs or firewalls

## Technical Specifications
- **Memory Footprint**:
  - Program memory: ~60 kB
  - Stack size: Minimum 5 kB (recommended)
  - Heap usage: ~2 kB on startup, ~4 kB per connected client
- **Connectivity**: TCP or TLS transport
- **API**: Simple C API with broker configuration structure
- **Dependencies**:
  - ESP-IDF components: newlib, esp_timer, esp_tls (for secure TLS connections)
  - Managed components: sock_utils (for porting)

> ## Implementation Details

Reviewer 2: Does this section focus on the implementation details of the MQTT broker itself or the implementation details used on the ESP32?

I have this question because you mention that `The ESP32 implementation uses the poll-based mechanism`. In this case, why also provide the overview of `mux layer` and `epoll` implementation?

Author: About both, but the `epoll` note is misleading, I'll remove it.

> The broker is implemented as a port of the original Mosquitto codebase. It follows a single-threaded, event-driven architecture that efficiently handles multiple connections through socket multiplexing. The core of this implementation consists of several key components:

Reviewer 2: ```suggestion
The broker is implemented as a port of the original Mosquitto codebase. It follows a single-threaded, event-driven architecture that efficiently handles multiple connections through socket multiplexing. The core of this implementation consists of a number of key components covered below.
```

> The broker is implemented as a port of the original Mosquitto codebase. It follows a single-threaded, event-driven architecture that efficiently handles multiple connections through socket multiplexing. The core of this implementation consists of a number of key components covered below.

Reviewer 2: ```suggestion
The MQTT broker is implemented as a port of the original Mosquitto codebase. It follows a single-threaded, event-driven architecture that efficiently handles multiple connections through socket multiplexing. The core of this implementation consists of a number of key components covered below.
```

### Socket Multiplexing
The broker uses an abstracted multiplexing layer that can utilize either poll-based or epoll-based implementations:
- **Mux Layer**: An abstraction layer (`mux.c`) that provides common interface functions for socket multiplexing

> - **Poll Implementation**: Uses standard poll() mechanisms for handling multiple file descriptors

Reviewer 2: Please use monospace to highlight all programming entities.

```suggestion
- **Poll Implementation**: Uses standard `poll()` mechanisms for handling multiple file descriptors
```

- **Epoll Implementation**: An optimized event notification mechanism (supported on some systems)

The ESP32 implementation uses the poll-based mechanism which is well-suited for the number of connections the ESP32 can realistically handle.

### Event Loop
The main broker loop (`loop.c`) performs the following operations:
1. Processes incoming client connections
2. Handles socket read/write events
3. Manages client keepalive tracking
4. Processes message queues and subscriptions
5. Handles disconnections and cleanup

### ESP-TLS Integration
The port includes a custom implementation (`net__esp_tls.c`) that integrates ESP-TLS for secure connections:
- Handles TLS context initialization
- Manages secure socket connections
- Processes TLS handshakes and secure data transfer

> - Maintains a fixed array of TLS contexts for connected clients

Reviewer 2: This part reads almost like technical documentation. It is an article advertizing the solution. It should have more sentences giving context and breaking things down for readers to have a better understanding.

Not sure what to do with this part. See if this information is needed in this article, and, if yes, try to come up with ideas on how to make this information more engaging.

Author: Simplified and reworded, +added links to ESP-TLS docs

### Simplified message handling
```mermaid
graph TD
    A[Client Connection Request] --> B[Socket Accept]
    B --> C{TLS Enabled?}
    C -->|Yes| D[ESP-TLS Context Creation]
    C -->|No| E[Plain TCP Connection]
    D --> F[TLS Handshake]
    F --> G[Connection Ready]
    E --> G
    G --> H[Socket Multiplexing]

    subgraph "Event Loop"
        H --> I[Poll for Events]
        I --> J[Process Socket Events]
        J --> K[Handle Client Messages]
        K --> L[Route Messages to Subscribers]
        L --> I
    end

    M[Client Disconnect] --> N[Socket Cleanup]
    N --> O[Release Resources]
```

> ## Getting Started

Reviewer 2: Quite likely, this section is the most important part of this article for readers.

Suggest making the title more explicit as to what they are getting started with.

How about this:

```suggestion
## How to use the MQTT broker in your project
```

> To use the ESP32 Mosquitto Port in your projects, you'll need to download the mosquitto component and initialize the broker with appropriate configuration settings and provide sufficient stack space for the running task.

Reviewer 2: Link to the mosquitto component?

Author: added

> To use the ESP32 Mosquitto Port in your projects, you'll need to download the mosquitto component and initialize the broker with appropriate configuration settings and provide sufficient stack space for the running task.

Reviewer 2: and is repeated twice. How about this:

```suggestion
To use the ESP32 Mosquitto Port in your projects, you'll need to download the mosquitto component and initialize the broker with appropriate configuration settings. It is important to provide sufficient stack space for the running task.
```

To use the ESP32 Mosquitto Port in your projects, you'll need to download the mosquitto component and initialize the broker with appropriate configuration settings and provide sufficient stack space for the running task.

### Start from an existing example
Use the below command to create a new project from the simple broker example
```
$ idf.py create-project-from-example "espressif/mosquitto:broker"
```

> Choose the preferred connection and example transport (TLS or plain TCP), build and flash normally with:

Reviewer 2: This is a getting started section.

I shouldn't live the readers wondering about to how certain things are done.

Regarding how to choose between TLS or plain TCP, either mention it here if it is easy or refer to the respective documentation.

Author: Added a new step, explaining `menuconfig` and the project configuration menu.

```
$ idf.py build flash monitor
```

> ### Use mosquitto in your project

Reviewer 2: Am I correct that this article focuses on the port of Mosquitto and not Mosquitto itself?

If yes, what do you want to call the port throughout the article: the MQTT broker?

Please decide what works better and make this term use consistent throughout the article.

Author: The article is primarily about Mosquitto port, but some sections (for example the tech details) about the Mosquitto itself.
Here it's about the "port" again, but the formal component name is just `"mosquitto"` (not `esp-mosquitto`).

Maybe, I'll just rephrase it to something neutral, like "Use the component in your project"?

> Choose the preferred connection in the "Connection" configuration menu and example transport (TLS or plain TCP) in the "Example" configuration menu.

Reviewer 2: Suggest making this paragraph more structured:

```suggestion
In the appeared configuration menu:

- Go to `Connection` and choose the preferred connection
- Go to `Example` and choose the example transport (TLS or plain TCP)
```

- Plugins and configuration files to extend broker functionality
- WebSocket and Websocket Secure transports

## License Information

The ESP32 Mosquitto Port is licensed under the dual Eclipse Distribution License (EDL) and Eclipse Public License (EPL). These licenses ensure that the software remains free and open-source while providing clear terms for usage, modification, and distribution.

- **Eclipse Distribution License (EDL)**: A BSD-style license that allows for commercial use, modification, distribution, and private use with minimal restrictions.
- **Eclipse Public License (EPL)**: A copyleft license that requires any distributed modifications to also be licensed under the EPL.

For full license details, please refer to the LICENSE files included in the source code repository:
- [edl-v10](https://github.com/espressif/esp-protocols/blob/mosq-v2.0.20_2/components/mosquitto/edl-v10) - Eclipse Distribution License version 1.0
- [epl-v20](https://github.com/espressif/esp-protocols/blob/mosq-v2.0.20_2/components/mosquitto/epl-v20) - Eclipse Public License version 2.0

### Trademark Attribution

> All Mosquitto logos, "Eclipse Mosquitto", and related trademarks are the property of the Eclipse Foundation. The use of these marks is subject to the Eclipse Foundation's trademark guidelines and does not imply any endorsement by the Eclipse Foundation.

Reviewer 2: Please correct me if I am wrong, but I believe there is no need for a separate section for trademarks.

In our recent article, we [provided the trademark attribution statement](https://developer.espressif.com/blog/2025/02/rainmaker-over-thread/#other-useful-links) in the article footer.

The [Eclipse Foundation Trademark Usage Policy](https://www.eclipse.org/legal/logo-guidelines/) also doesn't give any explicit instructions regarding that.

However, the trademark atribution statement should be a bit different according to the guidelines.

Suggest that you remove this section and include the following at the bottom of your article:

```html
<div style="font-size: 0.8em; color: #888; margin-top: 2em;">
Eclipse, Eclipse Mosquitto, and the Eclipse Foundation logos are trademarks of Eclipse Foundation, Inc.
</div>
```

Author: Thanks for the pointers, applied per your suggestion.
